lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php

```diff
@@ -290,10 +290,10 @@ protected function gatherRowData(array $data, array &$id, array &$nonemptyCompon
                     // in an inheritance hierarchy the same field could be defined several times.
                     // We overwrite this value so long we don't have a non-null value, that value we keep.
                     // Per definition it cannot be that a field is defined several times and has several values.
```

Comment:
This particular definition is what is being challenged by your inheritance design. The field definition should be fetched per entity name, and `$fieldName` should be specific to the currently hydrated entity type instead of just the name of a field. Did you try step-debugging here? What is `$type`?